src/graphs/hld.rs

```diff
@@ -0,0 +1,129 @@
+//! # Heavy Light Decomposition
+
+use crate::graphs::dfs_order::{get_dfs_postorder, get_dfs_preorder};
+use std::ops::Range;
+
+/// # Example
+/// ```
+/// use programming_team_code_rust::graphs::hld::HLD;
+/// use programming_team_code_rust::data_structures::fenwick::Fenwick;
+///
+/// let mut adj = vec![
+///    vec![1, 2],
+///    vec![0, 3, 4],
+///    vec![0, 5],
+///    vec![1],
+///    vec![1],
+///    vec![2],
+/// ];
+/// let mut fenwick = Fenwick::<usize>::build_on_array(&[3, 2, 4, 5, 1, 2]);
+///
+/// let hld = HLD::new(&mut adj, false);
+/// assert_eq!(hld.lca(3, 4), 1);
+/// assert_eq!(fenwick.sum(hld.sub_tree(0)), 17);
+/// let mut sum = 0;
+/// hld.path(3, 4, |range, _| sum += fenwick.sum(range));
+/// assert_eq!(sum, 11);
+/// ```
+pub struct HLD {
+    /// parent
+    pub p: Vec<usize>,
+    /// time in
+    pub tin: Vec<usize>,
+    siz: Vec<usize>,
+    head: Vec<usize>,
+    vals_edges: bool,
+}
+
+impl HLD {
+    /// Create a new HLD struct
+    ///
+    /// `adj` can be undirected tree or a directed tree (rooted at node 0)
+    /// `adj` can also be an undirected forest
```

Comment:
I didn't test it on forest (but yosupo is undirected, aizu is directed/rooted)

I felt since it depends on the dfs_order thing (which is already tested on forest), that it's okay to not test it here